twinkle-app/twinkle.bll.services/src/test/java/com/github/donkeyrit/twinkle/bll/UserRegisterTest.java

```diff
@@ -0,0 +1,96 @@
+import org.testng.Assert;
+import org.testng.annotations.*;
+
+public class UserRegisterTest extends BaseTest {
+
+    private UserRepository userRepository;
+    private User user;
+
+    @BeforeClass
+    public void setUp() {
+        EntityManager session = sessionFactory.createEntityManager();
+        userRepository = new UserRepositoryImpl(session);
+    }
+
+    @Test
+    void registerNewUser(){
+        // Arrange
+        user = Utils.generateTestUser();
+        String userPassword = user.getPassword();
+        user.setPassword(HashManager.generateHash(user.getPassword()));
+
+        // Act
+        LoginService loginService = new DefaultLoginService(userRepository);
+        AuthenticationResult authResult = loginService.signUp(user.getLogin(), userPassword, userPassword);
+
+        // Assert
+        Assert.assertNull(authResult.errorMessage());
+        Assert.assertTrue(authResult.isSuccessfull());
+    }
+
+    @Test
+    void registerExistenceUser(){
+        // Arrange
+        user = Utils.generateTestUser();
+        String userPassword = user.getPassword();
+        user.setPassword(HashManager.generateHash(user.getPassword()));
+        userRepository.insert(user);
+
+        // Act
+        LoginService loginService = new DefaultLoginService(userRepository);
+        AuthenticationResult authResult = loginService.signUp(user.getLogin(), userPassword, userPassword);
+
+        // Assert
+        Assert.assertEquals(authResult.errorMessage(), "Login already exist");
+        Assert.assertFalse(authResult.isSuccessfull());
+    }
+
+    @DataProvider(name = "userData")
+    public Object[][] provideUserData() {
+        return new Object[][] {
+                {"test_user_login", "test_user_password", ""},
+                {"test_user_login", "", "test_user_password"},
+                {"", "test_user_password", "test_user_password"},
+                {"test_user_login", "", ""},
+                {"", "test_user_password", ""},
+                {"", "", "test_user_password"},
+                {"", "", ""}
+        };
+    }
+
+    @Test(dataProvider = "userData")
+    void registerUserWithEmptyFields(String username, String password, String confirmPassword){
+        // Arrange
+        user = Utils.generateTestUser();
+        user.setPassword(HashManager.generateHash(user.getPassword()));
+
+        // Act
+        LoginService loginService = new DefaultLoginService(userRepository);
+        AuthenticationResult authResult = loginService.signUp(username, password, confirmPassword);
+
+        // Assert
+        Assert.assertEquals(authResult.errorMessage(), "All fields are required.");
+        Assert.assertFalse(authResult.isSuccessfull());
+    }
+
+    @AfterTest
+    void tearDown(){
+        if (user != null) {
+            System.out.println("Remove the test user: " + user.getLogin());
+            var temp = userRepository.getByLoginAndPassword(user.getLogin(), user.getPassword());
```

Comment:
You don't need to retrieve user again, just use object user from `AuthenticationResult`.

Reply:
Done